common/djangoapps/student/views.py

```diff
@@ -2099,6 +2083,50 @@ def create_account_with_params(request, params):
     return new_user
 
 
+def should_skip_activation_email(user, do_external_auth, running_pipeline, third_party_provider):
+    """
+    Return `True` if activation email should be skipped.
+
+    Skip email if we are:
+        1. Doing load testing.
+        2. Random user generation for other forms of testing.
+        3. External auth bypassing activation.
+        4. Have the platform configured to not require e-mail activation.
+        5. Registering a new user using a trusted third party provider (with skip_email_verification=True)
+
+    Note that this feature is only tested as a flag set one way or
+    the other for *new* systems. we need to be careful about
+    changing settings on a running system to make sure no users are
+    left in an inconsistent state (or doing a migration if they are).
+
+    Arguments:
+        user (User): Django User object for the current user.
+        do_external_auth (bool): True if external authentication is in progress.
+        running_pipeline (dict): Dictionary containing user and pipeline data for third party authentication.
+        third_party_provider (ProviderConfig): An instance of third party provider configuration.
+
+    Returns:
+        (bool): `True` if account activation email should be skipped, `False` if account activation email should be
+            sent.
+    """
+    # Get user's email returned by the third party authentication provider
+    user_email = running_pipeline and running_pipeline['kwargs'].get('details', {}).get('email')
+
+    skip_email = (
+        settings.FEATURES.get('SKIP_EMAIL_VALIDATION', None) or
+        settings.FEATURES.get('AUTOMATIC_AUTH_FOR_TESTING') or
+        (settings.FEATURES.get('BYPASS_ACTIVATION_EMAIL_FOR_EXTAUTH') and do_external_auth) or
+        (
+            third_party_provider and (
+                third_party_provider.skip_email_verification or
+                ((user.email != user_email) if user_email is not None else False)
+            )
+        )
+    )
+
+    return skip_email
```

Comment:
```
sso_pipeline_email = running_pipeline and running_pipeline['kwargs'].get('details', {}).get('email')

# Email is valid if the SAML assertion email matches the user account email or
# no email was provided in the SAML assertion. Some IdP's use a callback
# to retrieve additional user account information (including email) after the
# initial account creation.
valid_email = sso_pipeline_email == user.email or sso_pipeline_email is None

return (
    settings.FEATURES.get('SKIP_EMAIL_VALIDATION', None) or
    settings.FEATURES.get('AUTOMATIC_AUTH_FOR_TESTING') or
    (settings.FEATURES.get('BYPASS_ACTIVATION_EMAIL_FOR_EXTAUTH') and do_external_auth) or
    ( third_party_provider and third_party_provider.skip_email_verification and valid_email)
```

Reply:
The valid email check should actually be:
```
from common.djangoapps.third_party_auth.saml import SAP_SUCCESSFACTORS_SAML_KEY

valid_email = (
    sso_pipeline_email == user.email or (
        sso_pipeline_email is None and
        third_party_provider.identity_provider_type == SAP_SUCCESSFACTORS_SAML_KEY
    )
)
```

Reply:
`third_party_provider.identity_provider_type == SAP_SUCCESSFACTORS_SAML_KEY` check does not for SAML IdP that are not associated with SAP. 
With above check activation email is being sent to when `third_party_provider.skip_email_verification  == True` and I am signing in using testship IdP.
here is the sample `details` dict in the IdP response.
```
{
    u'username': u'myself', 
    u'fullname': u'Me Myself And I', 
    u'last_name': u'And I', 
    u'first_name': u'Me Myself', 
    u'email': None
}
```

Reply:
Sorry, I'm not following your last comment regarding the identity_provider_type check. I think we only want to skip the activation email when the SAML email address is None AND the IdP is SuccessFactors (or some other IdP that uses some kind of callback to get additional user profile information).

Reply:
I made the changes you request above

>```from common.djangoapps.third_party_auth.saml import >SAP_SUCCESSFACTORS_SAML_KEY
>valid_email = (
>    sso_pipeline_email == user.email or (
>        sso_pipeline_email is None and
>        third_party_provider.identity_provider_type == SAP_SUCCESSFACTORS_SAML_KEY
>    )
>)

but it was not skipping email (for testshib IdP) for the following reason
1. `sso_pipeline_email is None` returned `True` but `third_party_provider.identity_provider_type == SAP_SUCCESSFACTORS_SAML_KEY` returned `False`.

second condition failed because `third_party_provider.identity_provider_type` was `standard_saml_provider` instead of `sap_success_factors`.

So, I am guessing that there might be identity providers for which `third_party_provider.identity_provider_type == SAP_SUCCESSFACTORS_SAML_KEY` would fail and they would have missing email in user data, hence we should remove the above condition from the expression.

Reply:
That’s the point. We would not want to skip sending the activation email under the condition you describe in #1 in your last comment.

Reply:
Done